### ISSUE TITLE
Higher-order derivatives extremely slow, increasing exponentially

### ISSUE THREAD

**Issue by Acciorocketships** [pytorch/pytorch#111563]
### 🐛 Describe the bug

In my application, I need to take the nth order mixed derivative of a function. However, I found that the torch.autograd.grad computation time increases exponentially as n increases. Is this expected, and is there any way around it?

This is my code for differentiating a function self.F (from R^n -> R^1):

```
def differentiate(self, x):
    x.requires_grad_(True)
    xi = [x[...,i] for i in range(x.shape[-1])]
    dyi = self.F(torch.stack(xi, dim=-1))
    for i in range(self.dim):
        start_time = time.time()
        dyi = torch.autograd.grad(dyi.sum(), xi[i], retain_graph=True, create_graph=True)[0]
        grad_time = time.time() - start_time
        print(grad_time)
    return dyi
```

And these are the times printed for each iteration of the above loop:
```
0.0037012100219726562
0.005133152008056641
0.008165121078491211
0.019922733306884766
0.059255123138427734
0.1910409927368164
0.6340939998626709
2.1612229347229004
11.042078971862793
```

I assume this is because the size of the computation graph is increasing? Is there any way around this? I thought I might be able to circumvent this issue by taking a functional approach (presumably obviating the need for a computation graph), using torch.func.grad. However, this actually increased the runtime of the same code! Am I not understanding torch.func.grad properly?

### Versions

torch 2.1.0

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

**Comment by jon-chuang** [comment]
To my understanding, the number of gradients you have to compute and store, increases exponentially in number of params.

**Comment by Acciorocketships** [comment]
@jon-chuang @soulitzer 
Why is this the case? It must be more complicated than just `num_params ** order`. For example, if you have a polynomial with n terms, then you can easily compute the k-th derivative in linear time. I can see how a function like sin(x^2) would explode, as the first derivative is 2x*cos(x^2), the second is -4x*sin(x^2)+2*cos(x^2), so you end up getting more terms for each derivative you take, serving as a branching factor that leads to exponential time (also, i assume torch does not combine like terms to shrink the computation graph). In most people's cases, nonlinear functions like that only come into play in the activation function. If a layer is defined `f(w*x+b)` where f is the activation, then the derivative d/dx is `w*f'(w*x+b)`. w does not depend on x, so presumably the number of terms can only grow depending on your choice in activation function. Does this mean that the choice in activation function is crucial to the speed of higher-order derivatives, and some choices in activation functions can significantly reduce the runtime?

**Comment by jon-chuang** [comment]
I may have misunderstood your code. I see that you grad on the sum of the grads in previous level so you would nor have the naive exponential number of derivatives problem.

You may be right, I think the grad has at most constant factor more operators than forward, something like 3-4. And it seems that the successive execution times increase by 3-4 per order.

**Comment by Acciorocketships** [comment]
Sure, I just ran it with ReLU:

```
0.01187586784362793
0.00047016143798828125
0.00043773651123046875
0.0005078315734863281
0.00047707557678222656
0.00044083595275878906
0.00038886070251464844
0.00038313865661621094
0.0003960132598876953
```

**Comment by jon-chuang** [comment]
Interesting, it seems to stay approximately constant. I'd say problem solved? But if anyone could link a result from the literature on the constant factor overhead that would be much appreciated (been trying to scour the web for it to no avail)

**Comment by Acciorocketships** [comment]
It is interesting, but unfortunately Mish was just for testing. In my application I need to use a fairly complex activation function (the n-th integral of erf(x)).

This is more speculative, but would it be possible to significantly increase the computation time if we can provide an analytical solution to the nth derivative? I wonder if that would allow us to circumvent the "branching factor", where torch attempts to take the derivative of each term. I realise that the implementation would be fairly tricky, so I'm mostly asking if this is theoretically possible.

**Comment by jon-chuang** [comment]
I think you could try it. 

You can try implementing custom gradients for your function by following this guide: 
https://pytorch.org/docs/stable/notes/extending.func.html#extending-torch-func-with-autograd-function

As an example, you can see how Triton implements a custom kernel for MHA [here](https://github.com/openai/triton/blob/bdf464e4a8f80ad6bd6a7b470cb3d36efd61c8a2/python/triton/ops/flash_attention.py#L295)

**Comment by Acciorocketships** [comment]
Oh, that's clever! So if you implement backward() for every derivative up to n, providing the result as a function of another custom module, then the branching factor can be eliminated?

**Comment by jon-chuang** [comment]
Perhaps - it also depends how efficient your backward implementation is. You may also want to try approximations if it speeds things up.

**Comment by Acciorocketships** [comment]
The implementation of backward is not an issue, I already have the analytical form of all of the derivatives. I just wanted to confirm the logic that this would solve the "branching issue" that causes exponential growth in the grad time.

Thanks for your help!

**Comment by jon-chuang** [comment]
Cool, let us know how it goes. Would love to hear if you can achieve the speedup desired.

**Comment by Acciorocketships** [comment]
Okay, the results are in. These are the times with my custom activation function without any optimisations:
```
0.01146697998046875
0.005676984786987305
0.016211986541748047
0.04466104507446289
0.14060378074645996
0.4138617515563965
1.1940059661865234
3.6735780239105225
11.722836256027222
```
And these are the results when I implement a custom backward pass that can be called recursively:
```
0.01222991943359375
0.00305938720703125
0.005202054977416992
0.009228229522705078
0.01624774932861328
0.03122401237487793
0.10631084442138672
0.14510798454284668
0.35104799270629883
```
A significant speedup, but still not feasible for extremely high-order derivatives! I realised that there is a separate problem. While this solves the "branching factor" for each backward pass due to multiple terms in the derivative of the activation, it cannot solve the problem introduced by the chain rule: if there are multiple layers, then the chain rule is applied, so you get multiple terms depending on x. In the next pass, the chain rule is applied again, *and* the product rule is applied to the two terms you currently have. I only see two ways around this: 1) use only one nonlinearity in your network , or 2) use ReLU, which does not require the chain rule to be applied because it is linear. Unfortunately, in my application I must use this specific activation function, so if I  don't want an exponential increase in computation time for higher-order derivatives I have to use a two layer MLP with only one activation function. Here are the results for the same network with 2 layers instead of 4 (I confirmed that 3 layers also explodes exponentially):
```
0.010089874267578125
0.0007302761077880859
0.0004699230194091797
0.00039124488830566406
0.00028896331787109375
0.00026702880859375
0.0002627372741699219
0.0001480579376220703
0.0001418590545654297
```

**Comment by jon-chuang** [comment]
> In the next pass, the chain rule is applied again, and the product rule is applied to the two terms you currently have

Right, so basically, the grad itself creates a branching factor due to the chain rule. I don't see how this can be overcome.

**Comment by jon-chuang** [comment]
Hmm, I don't see why this is the case. If we use the backprop algorithm to calculate the gradient of higher order derivatives with respect to the parameters of the function:

$\eta_0(x) =  F_{W}(x)$

$\eta_k(x) = \sum_W \nabla_W \eta_{k-1}(x) = \sum_k \sum_{W_k} \frac{\partial \eta_{k-1}(x)}{\partial a_n}  \cdot (\prod_{k < j \le n} \frac{\partial a_j}{\partial o_j} \cdot \frac{\partial o_j}{\partial a_{j-1}} ) \cdot \frac{\partial a_k}{\partial W_k}$

$= \sum_k   \sum_{W_k} \frac{\partial \eta_{k-1}(x)}{\partial a_n}  \cdot (\prod_{k < j \le n} \frac{\partial a_j}{\partial o_j} \cdot W_j ) \cdot (\frac{\partial a_k}{\partial o_k} \otimes a_{k-1})$

$= \sum_k  \sum_{W_k} \bar a_k \cdot (\frac{\partial a_k}{\partial o_k} \otimes a_{k-1})$

where:
- $a_0 = x$
- $o_j  = W_j \cdot a_{j-1}$
- $n$ is the number of layers
- $a_j$ are the output of the activation functions given $o_j$.

And the adjoints $\bar a_k$, derived via chain rule from scalar output to input, defined recursively as 

$\bar a_k = \frac{\partial \eta(x)}{\partial a_k} =  \frac{\partial \eta(x)}{\partial a_{k+1}} \cdot \frac{\partial a_{k+1}}{\partial o_{k+1}}  \cdot \frac{\partial o_{k+1}}{\partial a_k}$
$= \bar a_{k+1} \cdot \frac{\partial a_{k+1}}{\partial o_{k+1}} \cdot W_k$

If you were to apply grad to the products directly, then yes, you would have an exponential gradients problem. However, backprop ought to be able to avoid this.

**Comment by Acciorocketships** [comment]
I do have a custom grad for each higher order derivative. If I only use the custom grad for the first derivative then the performance is much worse.

**Comment by jon-chuang** [comment]
Hmm, basically, I don't see why you would have exponentially increasing. Could you share a few examples of each derivative? I think the reason is that their derivatives might contributing to the branching factor.

**Comment by jon-chuang** [comment]
No, I think this is wrong - this is due to symbolic differentiation. Backprop will not run into this issue.

**Comment by Acciorocketships** [comment]
If that's the case then why are we seeing the problem empirically?

In backprop we will multiply our derivative in a given module by the upstream derivative. So, at the end of the first pass, we will have a product of the derivatives for each module. In the next iteration (for a higher-order derivative), won't we need to take the derivative of all of these multiplied terms, creating more terms with the product rule?

**Comment by jon-chuang** [comment]
No. As you can see from the equations I wrote down above, backprop caches intermediate results using something called the adjoint method. This means that only activation functions can create more operators in a simple MLP.

**Comment by jon-chuang** [comment]
Maybe this reason is that it actually uses forward-mode differentiation? 🤔

**Comment by jon-chuang** [comment]
You might try to use this instead to compute the backwards? https://pytorch.org/docs/stable/generated/torch.autograd.functional.vjp.html

**Comment by jon-chuang** [comment]
Oh wait, I just realized that if you compute the higher order VJP, its gonna be exponential too? First, you double the graph, then you take the VJP of that, doubling the graph again...

**Comment by Acciorocketships** [comment]
I would expect the VJP to be slightly slower than the current method because right now I am only computing the derivatives for one input, whereas with the VJP I would be computing the derivatives for all inputs and then multiplying by a one-hot (effectively slicing it). The size of the output should remain constant, but I don't see how it will solve the original problem.

**Comment by Acciorocketships** [comment]
I don't know about the adjoint method---what are the intermediate results that it caches? In the second-order derivative, how does it get around taking the derivative of each term in the product produced by the first-order derivative?

By the way, I just wanted to mention that second derivative (and all higher derivatives) of a network with ReLU is 0.

**Comment by jon-chuang** [comment]
Right. Anyway, regardless of using VJP or forward-mode, I suspect the reason is really that since you compute the higher order grad as the sum of the grads from the previous level, your graph gets longer and longer.

This is true whether you run forward-mode (due to chain rule) or reverse mode (due to doubling the graph when running backprop on each succesive graph)

**Comment by jon-chuang** [comment]
<img width="811" alt="image" src="https://github.com/pytorch/pytorch/assets/9093549/df963800-a7dc-4717-a1cf-101a32360c0f">

this is an example of a forwards + backwards graph.

**Comment by Acciorocketships** [comment]
The sum is only over the batch dimension, just like you would do when computing a loss. The different batch dimensions never interact, so it is not a branching factor. I can confirm this by running on a batch size of 1, where we see the same issue:
```
0.010080099105834961
0.0022132396697998047
0.003983259201049805
0.007370948791503906
0.014194965362548828
0.028818845748901367
0.05858922004699707
0.13225102424621582
0.3321499824523926
0.7825288772583008
```

**Comment by albanD** [comment]
Hey!

I would recommend as a nice tool to visualize how big your graph is https://github.com/szagoruyko/pytorchviz
You can add `        torchviz.make_dot(dyi.sum()).view()` to your for loop to see how the autograd graph evolves at every iteration.
Each xi is associated with one of the select_backward() node at the top.

But in this case, indeed the problem is that the double backward depends on the backward and thus you keep adding more and more to the graph (which you don't with relu as there is no dependency).

**Comment by soulitzer** [comment]
> In backprop we will multiply our derivative in a given module by the upstream derivative. So, at the end of the first pass, we will have a product of the derivatives for each module. In the next iteration (for a higher-order derivative), won't we need to take the derivative of all of these multiplied terms, creating more terms with the product rule?

@Acciorocketships  is right, the graph would basically double every time you apply chain rule + product rule.

It also does sound like techniques like wrapping hardcoded backward rules in custom Function would somewhat reduce the branching factor, though the degree that would help would depend on the specific function, and very higher order derivatives would still be problematic.

> No. As you can see from the equations I wrote down above, backprop caches intermediate results using something called the adjoint method. This means that only activation functions can create more operators in a simple MLP.

Yes there is some caching, because we save variables for backward, but that is not quite enough. There's actually a lot of recomputation going on when doing higher-order derivatives. If you are doing the nth derivative of something like `sin(sin(sin(` you are continually recomputing the 6 quantities (`cos(sin(sin`, `cos(sin`, `cos`, `sin(sin(sin`, `sin(sin`, `sin`), and it turns out you can do some more caching.

The below table is from doing the 7th derivative of `sin(sin(sin` on a (2048, 1024) tensor on cpu with and without caching. The tensor is large because we need to be somewhat compute-heavy, to benefit from caching.

| Condition         |  Time (s) | sin| sum| ones_like| expand| cos| mul| neg| add|
|-------------------|----------------|-------------|-------------|-------------------|----------------|-------------|------------|-------------|------------|
| Without Cache     | 6.186         | 149         | 7           | 7                 | 7              | 155         | 3270       | 478         | 2199       |
| With Cache        | 4.442          | 149         | 7           | 7                 | 7              | 155         | 3270       | 478         | 2199       |
| Pre-Dispatch Cache| 3.365             | 21          | 7           | 7                 | 7              | 20          | 2207       | 189         | 1726       |

<details>
<summary>
click to expand and see code
</summary>

```python
from torch.utils._python_dispatch import TorchDispatchMode
from torch.overrides import TorchFunctionMode
from torch.utils.weak import WeakTensorKeyDictionary
from torch._dispatch.python import enable_python_dispatcher, enable_pre_dispatch
from torch._C import DispatchKey
import time

torch.random.manual_seed(0)

ops_count = {}

class SinCosCache(TorchDispatchMode):
    def __init__(self, pre_dispatch=False):
        if pre_dispatch:
            super().__init__(DispatchKey.PreDispatch)
        self.cached_ops = (
            torch.ops.aten.sin.default,
            torch.ops.aten.cos.default,
        )
        self.caches = {f: WeakTensorKeyDictionary() for f in self.cached_ops}
        self.cache_hits = {f: 0 for f in self.cached_ops}
        self.cache_misses = {f: 0 for f in self.cached_ops}

    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
        ops_count[func] = ops_count.get(func, 0) + 1
        kwargs = {} if kwargs is None else kwargs
        if func in self.caches:
            cache = self.caches[func]
            if args[0] not in cache:
                cache[args[0]] = func(*args, **kwargs)
                self.cache_misses[func] += 1
            else:
                self.cache_hits[func] += 1
            return cache[args[0]]
        return func(*args, **kwargs)

# A fairly large input, and on cpu!
a = torch.rand(2048, 1024, dtype=torch.float64, requires_grad=True)

def fn(a):
    return a.sin()

def run():
    out = fn(fn(fn(a)))
    for i in range(7):
        print(i)
        out = torch.autograd.grad(out.sum(), a, create_graph=True)[0]

def do_time():
    start = time.time()
    run()
    end = time.time()
    return end - start

def print_ops_count():
    for op, count in ops_count.items():
        print(f"{op.__name__}: ", count)

with enable_python_dispatcher(), enable_pre_dispatch(), SinCosCache() as cache:
    print("cache: ", do_time())
print_ops_count()
ops_count = {}

print("no cache: ", do_time())

with enable_python_dispatcher(), enable_pre_dispatch(), SinCosCache(pre_dispatch=True) as cache:
    print("pre_dispatch cache: ", do_time())
print_ops_count()
```

</details>

In summary, I don't think there's not too much that can be done. Hardcoding higher-order rules to mitigate branching factor seems use case specific, and it seems that custom Function is a good enough extension point today to enable that.

It's possible that some kind of caching helper can be a useful tool for higher-order backprop in general, but the as with selective checkpointing, caching beneath autograd today is error prone due to version counter issues. It also does not seem to be impactful enough to impact very higher-order derivatives mainly because the runtime remains dominated by mul and mul cannot be easily cached because its very order dependent in that you have to multiply the terms in the exact same order to get cache hits on your muls. Thinking along those lines, the random idea would be to have a mechanism (perhaps a subclass) that performs muls lazily during backward until a non-linearity is reached at which point that mul would be materialized, so that maybe there would be more opportunities to hit a cache .

**Comment by soulitzer** [comment]
Leaving this issue as "needs research" since there is a lot of unnecessary computation with higher order derivatives, and there seem to be some general solutions that can help improve performance by caching etc. However, so far the improvements have only been demonstrated a very basic dummy model. Having more real life usecases would be useful to clarify how useful this would be in core.
